src/osd/PG.cc

```diff
@@ -7441,12 +7401,14 @@ PG::RecoveryState::Clean::Clean(my_context ctx)
   if (pg->info.last_complete != pg->info.last_update) {
     ceph_abort();
   }
-  pg->finish_recovery(*context< RecoveryMachine >().get_on_safe_context_list());
+  Context *c = pg->finish_recovery();
+  context< RecoveryMachine >().get_cur_transaction()->register_on_commit(c);
+  context< RecoveryMachine >().get_cur_transaction()->nop();
```

Comment:
What? Do we need an OP_NOP to ensure we commit metadata of some kind (and this got broken in the transaction refactor), or is this forcing us to disk in cases where we don't actually need to do so?

Reply:
because of this code in dispatch_transaction:
```
  if (ctx.transaction->empty() || !pg) {
    delete ctx.transaction;
  } else {
    int tr = store->queue_transaction(
      pg->ch,
      std::move(*ctx.transaction), TrackedOpRef(),
      handle);
    delete (ctx.transaction);
    assert(tr == 0);
  }
```

Reply:
How has that changed from the previous behavior, though?

Reply:
The old code checked the C_Contexts for empty too
```
-  if ((ctx.on_applied->empty() &&
-       ctx.on_safe->empty() &&
-       ctx.transaction->empty()) || !pg) {
```

Reply:
Yes, but before we didn't need to fill in a NOP. Now we are filling in a NOP. Do you contend there was a prior bug that you noticed and fixed here? Did some behavior change in a way I'm not seeing to require this NOP?

Reply:
Before simply adding the completion was sufficient to make it submit an empty transaction. Now it doesn't check for completions, on is_empty().  Alternative fixes include (1) making Transaction::empty() return false if there is a registered completion, or (2) making the dispatch_context code also check if the transaction has registered completions.  Or leaving this no-op.  I don't like (1) because it requires and audit of all other Transaction::empty() callers to see if they might be affected by the behavior change.  (2) would be fine with me

Reply:
(2) sounds good.